Initial Functionality Implementation

mcdonnnj opened the pull request:
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR is the initial implementation of the functionality of this Python package. That functionality is to retrieve an HTTP URL and hash the contents found, with additional processing for certain types of content.
<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

This is a core piece of the [BOD 20-01](https://cyber.dhs.gov/bod/20-01/) compliance checking we are tasked with providing.
<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

All of our [pre-commit](https://pre-commit.com/) hooks pass. I have implemented code coverage for almost the entire code base, and those tests are also passing.
<!-- How did you test your changes? How could someone else test this PR? -->

<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
* [x] Tests have been added and/or modified to cover the changes in this PR.
* [x] All new and existing tests pass.

mcdonnnj committed:
Initial changes after cloing repository

Swap out instances of skeleton-python-library and do some initial reference
updates

mcdonnnj committed:
Remove example files and put in framework for this module

Get rid of or move any of the example library's files and put in place some
files to implement for this package's functionality.

mcdonnnj committed:
Add utility methods for hashing

Add utility methods to get a desired hashing object and to update and get the
digest from a hashing object.

mcdonnnj committed:
Add NamedTuples to store expected results

mcdonnnj committed:
Implement a class to provide URL hashing functionality

The class has a primary method UrlHasher.hash_url() for functionality with
helper methods. The main content processing is done by _handle* methods that
are designed to handle a specific content-type as given by the remote server.
These methods will always return a HandlerResult to standardize them, and the
UrlHasher.hash_url() method will use the UrlResult named tuple to provide
access convenience while making the results immutable.

mcdonnnj committed:
Add initial testing for hash_http_content.hasher

Also fix a missing import and update __init__.py to expose the UrlHasher class.

mcdonnnj committed:
Update the README

Update the README to include information about this package's functionality.

mcdonnnj committed:
Add testing for a supplied Chromium binary

Add logic to the GitHub Actions workflow to pull down a zipped binary from the
https://github.com/adieuadieu/serverless-chrome project and extract it to a
specified location. The test_hasher.py file has been updated to add testing
using this binary instead of what is detected/used by pyppeteer automatically.
This will only take place if the binary is detected in the specified location.

mcdonnnj committed:
Consolidate handler definitions

Declare the dictionary of handlers upfront instead of adding them individually.

mcdonnnj committed:
Implement a command line interface to the package

mcdonnnj committed:
Add some debug logging to hash_http_content.hasher

Also correct the name of a test for the same file.

mcdonnnj committed:
Adjust fallback handler usage

If the contents appear to be text, use the plaintext handler as a fallback
instead.

mcdonnnj committed:
Add testing for the command line interface

Also adjust the output of the interface slightly.

mcdonnnj committed:
Add a script to retrieve serverless-chrome binaries

Switch the build workflow to use this script instead to retrieve it in CI.

jsf9k commented on setup.py:
```diff
@@ -42,16 +42,16 @@ def get_version(version_file):
+    download_url="https://github.com/cisagov/hash-http-content",
     # Author details
     author="Cyber and Infrastructure Security Agency",
     author_email="ncats@hq.dhs.gov",
```
This email address needs to be updated upstream.

mcdonnnj replied:
https://github.com/cisagov/skeleton-python-library/issues/57 Did we reach a consensus on which of each to use? I'm happy to make an upstream PR to resolve this.

dav3r replied:
Please see my comment at https://github.com/cisagov/skeleton-python-library/issues/57#issuecomment-776941344.

jsf9k commented on setup.py:
```diff
@@ -42,16 +42,16 @@ def get_version(version_file):
     long_description=readme(),
     long_description_content_type="text/markdown",
     # NCATS "homepage"
     url="https://www.us-cert.gov/resources/ncats",
```
This URL needs to be updated upstream.

jsf9k commented on src/hash_http_content/hasher.py:
```diff
@@ -0,0 +1,219 @@
+    # Not all implementations support the "usedforsecurity" keyword argument,
+    # which is used to indicate that the algorithm is being used for non-security
+    # related tasks. This is required for some algorithms on FIPS systems.
+    try:
+        hasher = getattr(hashlib, hash_algorithm)(usedforsecurity=False)
+    except AttributeError:
+        # There is no named constructor for the desired hashing algorithm
+        try:
+            # Work around typeshed's incorrect type hints
+            hasher = getattr(hashlib, "new")(hash_algorithm, usedforsecurity=False)
+        except TypeError:
+            hasher = hashlib.new(hash_algorithm)
+    except TypeError:
+        hasher = getattr(hashlib, hash_algorithm)()
+    return hasher
```
It's a complicated case...lotta ins, out, what-have-yous...

dav3r commented on README.md:
```diff
@@ -1,25 +1,39 @@
+[![Known Vulnerabilities](https://snyk.io/test/github/cisagov/hash-http-content/develop/badge.svg)](https://snyk.io/test/github/cisagov/hash-http-content)
+
+This is a Python library to retrieve the contents of a given HTTP URL and hash
+the processed contents.
```
```suggestion
This is a Python library to retrieve the contents of a given URL via HTTP (or
HTTPS) and hash the processed contents.
```

jsf9k commented on src/hash_http_content/hasher.py:
```diff
@@ -0,0 +1,219 @@
+    return hasher.hexdigest()
+
+
+class HandlerResult(NamedTuple):
```
Digging the named tuples!

jsf9k reviewed (approved):
Powerful work!

dav3r commented on src/hash_http_content/cli.py:
```diff
@@ -0,0 +1,105 @@
+"""Command line interface to the hash-url-content package.
```
```suggestion
"""Command line interface to the hash-http-content package.
```

dav3r commented on tests/test_hasher.py:
```diff
@@ -0,0 +1,311 @@
+
+def test_browser_additional_options():
+    """Verify that additional options are used in invoking the browser."""
+    # These options are expected for a lambda style environment
```
```suggestion
    # These options are expected for an AWS Lambda style environment
```

dav3r commented on tests/test_hasher.py:
```diff
@@ -0,0 +1,311 @@
+    if not os.path.isfile(serverless_chrome_path):
+        pytest.skip("no serverless-chrome binary found")
+
+    # These options are expected for a lambda style environment
```
```suggestion
    # These options are expected for an AWS Lambda style environment
```

dav3r commented on src/hash_http_content/hasher.py:
```diff
@@ -0,0 +1,219 @@
+    except AttributeError:
+        # There is no named constructor for the desired hashing algorithm
+        try:
+            # Work around typeshed's incorrect type hints
```
Can you include a comment with a link to more info or else just explain this a bit more?  I'm not familiar with what this refers to.

mcdonnnj replied:
Please see 5635e2a and let me know if that explains things more thoroughly.

dav3r replied:
Yes, thanks- that is very helpful!

dav3r commented on src/hash_http_content/hasher.py:
```diff
@@ -0,0 +1,219 @@
+    def hash_url(self, url: str) -> UrlResult:
+        """Get a hash of the contents of the provided URL."""
+        logging.debug("Hashing provided URL '%s'", url)
+        redirect_status_codes = [301, 307, 308]
```
I'm no expert on the intricacies of redirects, but there are other redirect codes besides these three- is there are reason you are excluding them (e.g. 302, 303, 305)?  https://tools.ietf.org/html/rfc7231#section-6.4

mcdonnnj replied:
I overlooked adding `302` when I was setting up that list, but my reasoning was to only count one type of redirect (per rfc7231):

> 1.  Redirects that indicate the resource might be available at a
       different URI, as provided by the Location field, as in the
       status codes 301 (Moved Permanently), 302 (Found), and 307
       (Temporary Redirect).

This is in line with the creation of `308` as described in https://tools.ietf.org/html/rfc7238.

jsf9k replied:
Excellent question and sharp eyes @dav3r!

@mcdonnnj, you should add the `302` code and add a comment explaining why you are only counting that one type of redirect.  That isn't immediately clear to me.

mcdonnnj replied:
I have added `302` and explanatory comments in a988651. Please let me know if there are any clarifications or further information you think it needs.

dav3r replied:
I'm good with your explanation in https://github.com/cisagov/hash-http-content/commit/a988651494d4639adc2d583a2d3095cc1eec369a - thanks!

dav3r reviewed (approved):
I left a few comments and questions, but overall, really great work @mcdonnnj! 🏎️

mcdonnnj committed:
Update comments based on feedback

Update comments with suggestions from code review.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Expand comment about mypy workaround

Add additional context and explanation about a workaround being used to pass
the mypy pre-commit hook.

jsf9k commented:
@cisagov/team-ois, we should give more specific comments when merging reviewer's `suggestion` blocks.  In this case, for example, "Update comments to explicitly mention AWS" would be a slightly better comment.  It gives you a little better idea of what is actually in the commit.

This PR doesn't contain a particularly egregious example, but I have seen `suggestion` commits in other PRs where the commit comment is "Modify blah.md", for example.  In such a case there is no way to know what the commit contains, short of examining the commit itself.  In many cases where a reviewer leaves behind multiple `suggestion` blocks it makes sense to group the suggestions into multiple commits that can each be given a specific commit message.

I'm not saying I always do it right either, but I've noticed that we sometimes merge others' suggestions without giving the commit comment the attention it deserves.  The comment should be just as informative as if you had created the commits yourself and were committing them from your editor or the command line.  Remember that when someone may be trying to figure out why things are done a certain way by studying the output of `git log` or `git whatchanged`, without the context of the PR.

mcdonnnj commented:
> 
> 
> @cisagov/team-ois, we should give more specific comments when merging reviewer's `suggestion` blocks. In this case, for example, "Update comments to explicitly mention AWS" would be a slightly better comment. It gives you a little better idea of what is actually in the commit.
> 
> This PR doesn't contain a particularly egregious example, but I have seen `suggestion` commits in other PRs where the commit comment is "Modify blah.md", for example. In such a case there is no way to know what the commit contains, short of examining the commit itself. In many cases where a reviewer leaves behind multiple `suggestion` blocks it makes sense to group the suggestions into multiple commits that can each be given a specific commit message.
> 
> I'm not saying I always do it right either, but I've noticed that we sometimes merge others' suggestions without giving the commit comment the attention it deserves. The comment should be just as informative as if you had created the commits yourself and were committing them from your editor or the command line. Remember that when someone may be trying to figure out why things are done a certain way by studying the output of `git log` or `git whatchanged`, without the context of the PR.

That's a really good point. Building on that, I think it's also important to only batch similar suggestions instead of just lumping everything together. In this PR it would have been helpful to do three commits from the suggestions instead of batching them all together. Then I could have done more specific commit messages for each.

mcdonnnj committed:
Add missing redirect status code and explain choices

The 302 status code was missed for the list of redirect_status_codes so it has
been added. I have also added an explanation for why these four values are
chosen instead of just matching any 3xx status code.

dav3r commented on src/hash_http_content/hasher.py:
```diff
@@ -0,0 +1,239 @@
+        logging.debug("Hashing provided URL '%s'", url)
+
+        # These values were chosen to keep in line with the type of redirection
+        # that indicates the desirted resource is at a different URI per
```
Fix typo:
```suggestion
        # that indicates the desired resource is at a different URI per
```

mcdonnnj committed:
Fix typo in comment

Fix a typo I made in the comment explaining the redirect statuses being used.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj committed:
Add content_type member to the UrlResult NamedTuple

Add support for the new member to the cli and its output. Update testing to
reflect this addition.

mcdonnnj committed:
Add a timeout instance variable to UrlHasher

This timeout is used for request.get() calls to provide a changeable limit for
how long to wait for a request to finish. The default value is five seconds.

mcdonnnj committed:
Update UrlHasher._handle_html() to navigate to a local file

When testing against different sites, I found that for more complex sites there
was not always enough time for everything to process and render. Unfortunately
the setContent() coroutine does not allow you to pass additional options. I
have switched to using the goto() coroutine instead, as this allows me ot use
the waitUntil option to give time for the page to process and fully load.

mcdonnnj committed:
Switch to reusing a browser page

Previously a new page was created, used, and then closed for every
UrlHasher_handle_html() call. However, since we already reuse the browser
object for a class instance, there's no sense in not reusing a page in the
browser in the same way.

mcdonnnj committed:
Handle browser timeout while waiting for content

Adjust the browser's timeout while waiting for the events given in the
waitUntil to occur from the default of 30 seconds to five seconds. Wrap
everything in a try/except block and pull whatever content is rendered at the
end of the timeout as a fallback. If a site has long-polling or similar side
activity, then the networkidle2 event may never occur. Handling the timeout
will allow us to retrieve a site that has likely loaded already.

mcdonnnj committed:
Add retry mechanism to requests.get() call

Add retries to the requests.get() call for the provided URL in
UrlHasher.hash_url(). This will allow a modified number of retries in case
there are any network difficulties or similar. After self._retries additional
attempts, it will raise the exception that was caught.

mcdonnnj committed:
Make Page.goto() timeout configurable

Change from a hard coded value to reusing the self._timeout value.

mcdonnnj committed:
Add option to control TLS validation

Added a verify kwarg to UrlHasher.hash_url() to allow control over TLS
verification in the requests library. The kwarg is directly used as the kwarg
of the same name in the requests.get() method.

mcdonnnj committed:
Narrow scope of try block in UrlHasher.hash_url()

When performing requests.get() we were blanket catching any Exception. Retries
should only be for connection related issued, so this narrows the caught
exceptions down to ConnectionError and Timeout (which will match both the
ConnectTimeout and ReadTimeout errors).

mcdonnnj committed:
Add hasher.UrlResult to the public objects

Since the result of a UrlHasher.hash_url() call and the UrlHasher class itself
are the primary interfaces to this class, they should both be publicly
available.

jsf9k commented on src/hash_http_content/hasher.py:
```diff
@@ -168,18 +169,37 @@ def _handle_html(self, contents: bytes, encoding: str) -> HandlerResult:
-        page: Page = asyncio.get_event_loop().run_until_complete(
-            self._browser.newPage()
+        # Try to guarantee our preferred encoding
+        page_contents = bytes(page_contents.encode(self._default_encoding)).decode(
```
Should this line have a type hint?

mcdonnnj replied:
Certainly wouldn't hurt. I get bad about it sometimes when a variable is only assigned to once from a "guaranteed output". I'll go back and add some more type hinting to keep it all fresh and statically happy (especially since I added new instance variables with the expectation that they can be changed).

mcdonnnj replied:
@jsf9k That is actually hinted here (just above the line you were looking at):
https://github.com/cisagov/hash-http-content/blob/2c51f2b29dc9971127572d86dc089a5947a6d5f9/src/hash_http_content/hasher.py#L216
, but I took care of some low-hanging/should be there type hints in 2c51f2b.

jsf9k reviewed (approved):
Now I like it even better!

mcdonnnj committed:
Add additional type hints to the UrlHasher class

During review type hints were mentioned, and it made me take a second look to
see if I had missed any that could be easily added. This was especially
important for some of the instance variables that were added, as they are
expected to be modified if desired.

dav3r reviewed (approved):
Nice improvements since the last time I looked! 👍
